libgeopmd/src/StatsCollector.cpp

```diff
@@ -0,0 +1,194 @@
+            result << "    " << "count: " << m_stats->count(metric_idx) << "\n";
+            result << "    " << "first: " << m_stats->first(metric_idx) << "\n";
+            result << "    " << "last: " << m_stats->last(metric_idx) << "\n";
+            result << "    " << "min: " << m_stats->min(metric_idx) << "\n";
+            result << "    " << "max: " << m_stats->max(metric_idx) << "\n";
+            result << "    " << "mean: " << m_stats->mean(metric_idx) << "\n";
+            result << "    " << "std: " << m_stats->std(metric_idx) << "\n";
+            ++metric_idx;
+        }
+        return result.str();
+    }
+
+    void StatsCollector::reset(void)
+    {
+        m_time_begin = geopm::time_curr_string();
+        m_stats->reset();
+    }
+
+}
+
+int geopm_stats_collector_create(size_t num_requests, const struct geopm_request_s *requests,
+                                 struct geopm_stats_collector_s **collector)
+{
+    int err = 0;
+    try {
+        std::vector<geopm_request_s> request_vec(requests, requests + num_requests);
+        auto result = std::make_unique<geopm::StatsCollector>(request_vec);
+        *collector = reinterpret_cast<geopm_stats_collector_s *>(result.release());
+    }
+    catch (...) {
+        err = geopm::exception_handler(std::current_exception());
+        err = err < 0 ? err : GEOPM_ERROR_RUNTIME;
+    }
+    return err;
+}
+
+int geopm_stats_collector_update(struct geopm_stats_collector_s *collector)
+{
+    int err = 0;
+    try {
+        geopm::StatsCollector *collector_cpp = reinterpret_cast<geopm::StatsCollector *>(collector);
+        collector_cpp->update();
+    }
+    catch (...) {
+        err = geopm::exception_handler(std::current_exception());
+        err = err < 0 ? err : GEOPM_ERROR_RUNTIME;
+    }
+    return err;
+}
+
+// If *max_report_size is zero, update it with the required size for the report
+// and do not modify report
+int geopm_stats_collector_report_yaml(const struct geopm_stats_collector_s *collector,
+                                      size_t *max_report_size, char *report_yaml)
+{
+    int err = 0;
+    try {
+        const geopm::StatsCollector *collector_cpp = reinterpret_cast<const geopm::StatsCollector *>(collector);
+        std::string report_str = collector_cpp->report_yaml();
+        if (report_str.size() < *max_report_size) {
+            strncpy(report_yaml, report_str.c_str(), report_str.size());
+        }
+        else {
+            err = ENOBUFS;
+            std::ostringstream err_str;
+            size_t report_str_size = report_str.size() + 1;
+            if (*max_report_size != 0) {
+                err_str << "geopm_stats_collector_report_yaml(): max_report_size is too small, provided: "
+                        << *max_report_size << " required: " << report_str_size;
+            }
+            else {
+                err_str << "geopm_stats_collector_report_yaml(): max_report_size required: " << report_str_size;
```

Comment:
Q: why treat this as an error if the interface spec says this function can be called with a size value of zero to query the actual size required? Returning an error to the caller might let them think the query was unsuccessful.

Reply:
My thinking is that success is determined by the report being fully copied into the buffer.  If the intent is to populate the `max_report_size`` value, the return code can be ignored.  On the other hand, if a user has submitted a buffer of zero length due to a bug, but the intent is to populate the buffer, the return code should be non-zero so the user is alerted to their bug.